first-read-last-write-cache/src/cache.rs

```diff
@@ -427,4 +428,42 @@ mod tests {
             }
         }
     }
+
+    proptest! {
+        #[test]
+        fn merge_doesnt_crash(a: u8, b: u8) {
+
+            let test_cases = vec![
+                TestCase {
+                    left: Some(ReadWrite::Read(new_cache_entry(a, b))),
+                    right: Some(ReadWrite::Read(new_cache_entry(a, b))),
+                },
+                TestCase {
+                    left: Some(ReadWrite::Read(new_cache_entry(a, b))),
+                    right: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                },
+                TestCase {
+                    left: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                    right: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                },
+                TestCase {
+                    left: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                    right: None,
+                },
+                TestCase {
+                    left: None,
+                    right: Some(ReadWrite::Read(new_cache_entry(a, b))),
+                },
+                TestCase {
+                    left: None,
+                    right: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                },
+                TestCase {
+                    left: Some(ReadWrite::Write(new_cache_entry(a, b))),
+                    right: Some(ReadWrite::Read(new_cache_entry(a, b))),
+                },
+            ];
+
```

Comment:
This only generates test cases but doesn't test the `merge` method. You can check the `test_merge_fail` test for details. I suggest extracting the code:
```
for TestCase { left, right } in test_cases.clone() {
            match (left, right) {
                (None, None) => {}
                (None, Some(rw)) => right_cache.add_to_cache(rw).unwrap(),
                (Some(rw), None) => left_cache.add_to_cache(rw).unwrap(),
                (Some(left_rw), Some(right_rw)) => {
                    left_cache.add_to_cache(left_rw).unwrap();
                    right_cache.add_to_cache(right_rw).unwrap();
                }
            }
        }

        let merged = left_cache.merge(right_cache).unwrap();
```
into a function and using it in all the `test_merge` tests.